// @iobroker/adapter-react/Components/Logo.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _Fab = _interopRequireDefault(require("@material-ui/core/Fab"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _i18n = _interopRequireDefault(require("../i18n"));

var _Help = _interopRequireDefault(require("@material-ui/icons/Help"));

var _VerticalAlignTop = _interopRequireDefault(require("@material-ui/icons/VerticalAlignTop"));

var _VerticalAlignBottom = _interopRequireDefault(require("@material-ui/icons/VerticalAlignBottom"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var styles = function styles(theme) {
  return {
    buttons: {
      marginRight: 5,
      marginTop: 5,
      "float": 'right'
    },
    logo: {
      padding: 8,
      width: 64
    }
  };
};
/**
 * @typedef {object} LogoProps
 * @property {any} common Adapter common configuration from io-package.json
 * @property {any} native Adapter native data from io-package.json
 * @property {number} instance Adapter instance number.
 * @property {(contents: any) => void} [onLoad]
 * @property {(error: string) => void} [onError]
 * @property {{ buttons: string, logo: string }} classes The styling class names.
 *
 * @extends {React.Component<LogoProps>}
 */


var Logo = /*#__PURE__*/function (_React$Component) {
  _inherits(Logo, _React$Component);

  var _super = _createSuper(Logo);

  function Logo() {
    _classCallCheck(this, Logo);

    return _super.apply(this, arguments);
  }

  _createClass(Logo, [{
    key: "handleFileSelect",
    value: function handleFileSelect(evt) {
      var _this = this;

      var f = evt.target.files[0];

      if (f) {
        var r = new window.FileReader();

        r.onload = function (e) {
          var contents = e.target.result;

          try {
            var json = JSON.parse(contents);

            if (json["native"] && json.common) {
              if (json.common.name !== _this.props.common.name) {
                _this.props.onError(_i18n["default"].t('ra_otherConfig', json.common.name));
              } else {
                _this.props.onLoad(json["native"]);
              }
            } else {
              _this.props.onError(_i18n["default"].t('ra_invalidConfig'));
            }
          } catch (e) {
            _this.props.onError(e.toString());
          }
        };

        r.readAsText(f);
      } else {
        alert('Failed to open JSON File');
      }
    }
  }, {
    key: "download",
    value: function download() {
      var result = {
        _id: 'system.adapter.' + this.props.common.name + '.' + this.props.instance,
        common: JSON.parse(JSON.stringify(this.props.common)),
        "native": this.props["native"]
      }; // remove unimportant information

      if (result.common.news) {
        delete result.common.news;
      }

      if (result.common.titleLang) {
        delete result.common.titleLang;
      }

      if (result.common.desc) {
        delete result.common.desc;
      } //window.open('data:application/iobroker; content-disposition=attachment; filename=' + result._id + '.json,' + JSON.stringify(result, null, 2));


      Logo.generateFile(result._id + '.json', result);
    }
  }, {
    key: "upload",
    value: function upload() {
      var _this2 = this;

      var input = window.document.createElement('input');
      input.setAttribute('type', 'file');
      input.setAttribute('id', 'files');
      input.setAttribute('opacity', 0);
      input.addEventListener('change', function (e) {
        return _this2.handleFileSelect(e, function () {});
      }, false);
      input.click();
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      return /*#__PURE__*/_react["default"].createElement("div", {
        className: this.props.className,
        style: this.props.style
      }, this.props.common.icon ? /*#__PURE__*/_react["default"].createElement("img", {
        src: this.props.common.icon,
        className: this.props.classes.logo,
        alt: "logo"
      }) : null, this.props.common.readme ? /*#__PURE__*/_react["default"].createElement(_Fab["default"], {
        size: "small",
        color: "primary",
        "aria-label": "Help",
        className: this.props.classes.buttons,
        onClick: function onClick() {
          var win = window.open(_this3.props.common.readme, '_blank');
          win.focus();
        }
      }, /*#__PURE__*/_react["default"].createElement(_Help["default"], null)) : null, /*#__PURE__*/_react["default"].createElement(_Fab["default"], {
        size: "small",
        color: "primary",
        "aria-label": "Load config",
        className: this.props.classes.buttons,
        title: _i18n["default"].t('ra_Load configuration from file'),
        onClick: function onClick() {
          return _this3.upload();
        }
      }, /*#__PURE__*/_react["default"].createElement(_VerticalAlignTop["default"], null)), /*#__PURE__*/_react["default"].createElement(_Fab["default"], {
        size: "small",
        color: "primary",
        "aria-label": "Save config",
        className: this.props.classes.buttons,
        title: _i18n["default"].t('ra_Save configuration to file'),
        onClick: function onClick() {
          return _this3.download();
        }
      }, /*#__PURE__*/_react["default"].createElement(_VerticalAlignBottom["default"], null)));
    }
  }], [{
    key: "generateFile",
    value: function generateFile(filename, obj) {
      var el = window.document.createElement('a');
      el.setAttribute('href', 'data:application/json;charset=utf-8,' + encodeURIComponent(JSON.stringify(obj, null, 2)));
      el.setAttribute('download', filename);
      el.style.display = 'none';
      window.document.body.appendChild(el);
      el.click();
      window.document.body.removeChild(el);
    }
  }]);

  return Logo;
}(_react["default"].Component);

Logo.propTypes = {
  common: _propTypes["default"].object.isRequired,
  className: _propTypes["default"].string,
  style: _propTypes["default"].object,
  "native": _propTypes["default"].object.isRequired,
  instance: _propTypes["default"].number.isRequired,
  onError: _propTypes["default"].func,
  onLoad: _propTypes["default"].func
};
/** @type {typeof Logo} */

var _export = (0, _withStyles["default"])(styles)(Logo);

var _default = _export;
exports["default"] = _default;
